### PR TITLE
School autocomplete progressive enhancement

### PR DESCRIPTION
**Pull request by benilovj**
### Context

Support agents need to assign users to new schools.

### Changes proposed in this pull request

- Add a JSON endpoint which returns schools based on a partial name or URN match
- Progressively enhance the school search box with an accessible autocomplete which uses the newly introduced JSON endpoint.

### Guidance to review

https://user-images.githubusercontent.com/23801/103418327-10884900-4b86-11eb-97ba-e992fa8c18c7.mp4

### CODE REVIEW

File: app/controllers/support/schools_controller.rb [hunk 1]
```diff
@@ -6,8 +6,18 @@ def search
   end
 
   def results
-    @search_form = BulkUrnSearchForm.new(search_params)
-    @schools = policy_scope(@search_form.schools).includes(:preorder_information, :responsible_body)
+    if request.post?
+      @search_form = BulkUrnSearchForm.new(search_params)
+      @schools = policy_scope(@search_form.schools).includes(:preorder_information, :responsible_body)
+    elsif request.get?
+      @form = Support::NewUserSchoolForm.new(name_or_urn: params[:query])
```
Comment (aldavidson) [inline app/controllers/support/schools_controller.rb:13]: Not a blocker, but this feels a bit fragile - and also limits potential for using the endpoint from other places

Comment (benilovj) [reply app/controllers/support/schools_controller.rb]: Agreed – the name of the form object isn't really reflective of its responsibilities so I've renamed it in f9e00d1

Comment (aldavidson) [reply app/controllers/support/schools_controller.rb]: :+1:

**Comment by benilovj** [comment]
@aldavidson I've rebased this PR so it no longer conflicts with `main`. Would you mind taking another look?

**Comment by aldavidson** [comment]
> @aldavidson I've rebased this PR so it no longer conflicts with `main`. Would you mind taking another look?

I've had a look, and tried it locally. 
The autocomplete works, and the original workflow is still there if you disable javascript - all good.

But I did find it a bit counter-intuitive that when you select a school from the autocomplete suggestion, you still then have to click 'Continue' to search for it, and then click 'Associate' to associate the school - I would have expected the act of clicking in the autocomplete list to replace at least the first of those clicks. I found myself repeatedly thinking I'd associated a school when I actually hadn't yet, and it feels like it adds a step to the workflow rather than removes one. Does that make sense?

**Comment by benilovj** [comment]
> > @aldavidson I've rebased this PR so it no longer conflicts with `main`. Would you mind taking another look?
> 
> I've had a look, and tried it locally.
> The autocomplete works, and the original workflow is still there if you disable javascript - all good.
> 
> But I did find it a bit counter-intuitive that when you select a school from the autocomplete suggestion, you still then have to click 'Continue' to search for it, and then click 'Associate' to associate the school - I would have expected the act of clicking in the autocomplete list to replace at least the first of those clicks. I found myself repeatedly thinking I'd associated a school when I actually hadn't yet, and it feels like it adds a step to the workflow rather than removes one. Does that make sense?

Yes this makes sense, but I don't want to make this PR bigger than it already is, and I have a follow-up in flight already that would change the interaction design (https://github.com/DFE-Digital/get-help-with-tech/pull/1000). Can I do the work you've suggested on the other PR?

**Comment by aldavidson** [comment]
> > > @aldavidson I've rebased this PR so it no longer conflicts with `main`. Would you mind taking another look?
> > 
> > 
> > I've had a look, and tried it locally.
> > The autocomplete works, and the original workflow is still there if you disable javascript - all good.
> > But I did find it a bit counter-intuitive that when you select a school from the autocomplete suggestion, you still then have to click 'Continue' to search for it, and then click 'Associate' to associate the school - I would have expected the act of clicking in the autocomplete list to replace at least the first of those clicks. I found myself repeatedly thinking I'd associated a school when I actually hadn't yet, and it feels like it adds a step to the workflow rather than removes one. Does that make sense?
> 
> Yes this makes sense, but I don't want to make this PR bigger than it already is, and I have a follow-up in flight already that would change the interaction design (#1000). Can I do the work you've suggested on the other PR?

Sure :+1:

**Review by aldavidson — APPROVED** [review]

**Pull request merged**